Source/WebKit/UIProcess/ApplicationStateTracker.h

```diff
@@ -83,10 +83,10 @@ class ApplicationStateTracker : public CanMakeWeakPtr<ApplicationStateTracker> {
 
     bool m_isInBackground;
 
-    id m_didEnterBackgroundObserver;
-    id m_willEnterForegroundObserver;
-    id m_willBeginSnapshotSequenceObserver;
-    id m_didCompleteSnapshotSequenceObserver;
+    RetainPtr<NSObject> m_didEnterBackgroundObserver;
+    RetainPtr<NSObject> m_willEnterForegroundObserver;
+    RetainPtr<NSObject> m_willBeginSnapshotSequenceObserver;
+    RetainPtr<NSObject> m_didCompleteSnapshotSequenceObserver;
```

Comment:
Generally speaking, it's not a good idea to retain observer objects because that has the potential to create a retain cycle with `this` (C++) or `self` (Objective-C).

Also, if the client (third-party) software wants to deallocate the observer object sooner, this object keeping them alive may cause other issues.

So I would suggest using `WeakObjCPtr<>` (see lines 70-73 above) or adding `__weak` to these instance variables so that you keep a weak reference to them.

If you use `WeakObjCPtr<>`, then the rest of the patch doesn't need to change.

If you use `__weak`, then you just need to get rid of the `.get()` calls below.

Reply:
I thought about this some more, and I think `RetainPtr<>` is okay if `ApplicationStateTracker` is not a (super) long-lived.

The reason objects like `NSNotificationCenter` don't take strong references is that if registered observers aren't removed, they'll basically leak for the life of the process.

So using either a weak or a strong reference is fine, though I'd still lean towards using a weak reference if possible (since it has the same behavior as a raw pointer, but without the downside of stale pointers when the observer object is deallocated but not unregistered).

Reply:
I'm testing with the `WeakObjCPtr<>` approach, and will very likely land that version.